databroker/broker.py

```diff
@@ -0,0 +1,448 @@
+    raise ValueError("Must give an integer scan ID like [6], a slice "
+                     "into past scans like [-5], [-5:], or [-5:-9:2], "
+                     "a list like [1, 7, 13], a (partial) uid "
+                     "like ['a23jslk'] or a full uid like "
+                     "['f26efc1d-8263-46c8-a560-7bf73d2786e1'].")
+
+
+@search.register(slice)
+def _(key, mds):
+    # Interpret key as a slice into previous scans.
+    logger.info('Interpreting key = %s as a slice' % key)
+    if key.start is not None and key.start > -1:
+        raise ValueError("slice.start must be negative. You gave me "
+                         "key=%s The offending part is key.start=%s"
+                         % (key, key.start))
+    if key.stop is not None and key.stop > 0:
+        raise ValueError("slice.stop must be <= 0. You gave me key=%s. "
+                         "The offending part is key.stop = %s"
+                         % (key, key.stop))
+    if key.stop is not None:
+        stop = -key.stop
+    else:
+        stop = None
+    if key.start is None:
+        raise ValueError("slice.start cannot be None because we do not "
+                         "support slicing infinitely into the past; "
+                         "the size of the result is non-deterministic "
+                         "and could become too large.")
+    start = -key.start
+    result = list(mds.find_last(start))[stop::key.step]
+    header = [Header.from_run_start(mds, h) for h in result]
+    return header
+
+
+@search.register(numbers.Integral)
+def _(key, mds):
+    logger.info('Interpreting key = %s as an integer' % key)
+    if key > -1:
+        # Interpret key as a scan_id.
+        gen = mds.find_run_starts(scan_id=key)
+        try:
+            result = next(gen)  # most recent match
+        except StopIteration:
+            raise ValueError("No such run found for key=%s which is "
+                             "being interpreted as a scan id." % key)
+        header = Header.from_run_start(mds, result)
+    else:
+        # Interpret key as the Nth last scan.
+        gen = mds.find_last(-key)
+        for i in range(-key):
+            try:
+                result = next(gen)
+            except StopIteration:
+                raise IndexError(
+                    "There are only {0} runs.".format(i))
+        header = Header.from_run_start(mds, result)
+    return header
+
+
+@search.register(str)
+@search.register(six.text_type)
+# py2: six.string_types = (basestring,)
+# py3: six.string_types = (str,)
+# so we need to just grab the only element out of this
+@search.register(six.string_types,)
+def _(key, mds):
+    logger.info('Interpreting key = %s as a str' % key)
+    results = None
+    if len(key) == 36:
+        # Interpret key as a complete uid.
+        # (Try this first, for performance.)
+        logger.debug('Treating %s as a full uuid' % key)
+        results = list(mds.find_run_starts(uid=key))
```

Comment:
I think we should leave this as is until https://github.com/NSLS-II/metadatastore/issues/205 is improved and MDS is subsequently tagged.